### ISSUE TITLE
How to implement in skorch a more complex early stopping criterion than a scalar-based one: combining F1/balanced accuracy and validation loss?

### ISSUE THREAD

**Issue by pdimitrov-thoughtriver** [skorch-dev/skorch#678]
Hello,

I've got the following issue which seems impossible to resolve with the current implementation of skorch. Namely, I use F1 or balanced accuracy for early stopping while training a model across many different yet related classification datasets, but for one particular classification dataset I reach the optimal value of 1 quite early in the training process while the validation loss continues to decrease for more iterations. Using the former - F1/accuracy - criterions leads to quite a suboptimal performance on a separate test set, while the validation loss optimal point produces much better results. Since, I want to keep F1/accuracy for the remaining datasets, I'd like to implement a version of the early stopping using the vector/tuple 
of F1/accuracy and validation loss, which will order improvements based on F1/accuracy first, followed by validation loss, only if F1/accuracy == 1.

For example:

F1      valid_loss     cp
.8       5.76               *
.9       3.51               *
.95     3.11               *
1.0      2.1                 *       <-- current checkpoint/best model
1.0      1.8
1.0      1.4                         <--- desired checkpoint/best model
1.0      2.5
.95     3.6

with F1 only model leading to:

Confusion matrix:
 [[1408    3]
 [  29   12]]

#positive = 12 out of 41, #negative = 1408 out of 1411
balanced accuracy = 64.53, balanced adjusted accuracy = 29.06
binary:	precision = 80.0, recall = 29.27, F1 = 42.86
macro:	precision = 88.99, recall = 64.53, F1 = 70.87

while "optimal validation loss following optimal F1"-based one leads to:

Confusion matrix:
 [[1408    3]
 [   5   36]]

#positive = 36 out of 41, #negative = 1408 out of 1411
balanced accuracy = 93.8, balanced adjusted accuracy = 87.59
binary:	precision = 92.31, recall = 87.8, F1 = 90.0
macro:	precision = 95.98, recall = 93.8, F1 = 94.86

Note that thresholds/everything else can stay the same since they can be used separately on the individual criterions, its allowing to have vector/tuple performance metric rather than scalar that is an issue.

What is the simplest possible solution that one can implement/hack quickly? I've quite an extensive modelling solution fully implemented in skorch, and so far all the problems that I had to address to develop my skorch solution were easily addressed, so big thanks for the great work are in order!

Thank you,

**Comment by BenjaminBossan** [comment]
Hi @pdimitrov-thoughtriver, sorry for the late reply.

If I understand your problem correctly, you would like to have a stopping criterion that chooses the best F1 and, in case of draws with F1=1, chooses the best valid loss.

For that, I would implement a completely new scoring function that mirrors exactly this logic. Here is some code to give you an idea:

```python
def my_score(net, X, y):
    # assuming that f1 and valid loss have already been computed
    f1 = net.history[-1, 'f1_score']
    if f1 < 1:
        return f1
    valid_loss = net.history[-1, 'valid_loss']
    return SOME_LARGE_NUMBER - valid_loss

...
net = NeuralNet(..., callbacks=[
    ...
    ('my_score', EpochScoring(my_score, lower_is_better=False, name='f1_or_loss')),
    ('stopping', EarlyStopping('f1_or_loss', lower_is_better=False)),
])
```

So the idea is that as long as F1 < 1, just return F1. But as soon as F1 == 1, you return SOME_LARGE_NUMBER - valid_loss. SOME_LARGE_NUMBER should be large enough that SOME_LARGE_NUMBER - valid_loss is always larger than 1, hence this value will always be bigger than pure F1. And we take -valid_loss because we want to maximize this new score. Does that make sense to you?

**Comment by pdimitrov-thoughtriver** [comment]
Hi @BenjaminBossan, no problem with the late reply and thanks for the suggestion. I ended up using a different approach, encoded as follows in this Python (pseudo-)code:

    def f1(net, X, y, alpha = 1e-5):
        return (
            max(net._score(X, y, score_type = "F1") - alpha, 0)
            + np.exp(-net.history[-1, 'valid_loss']) * alpha
        )


Basically, given a very small constant - called alpha - ideally, smaller than __any difference__ in F1 value across the whole validation set (as it varies between the assignments of Yes/No labels), the effect of the validation loss is encoded in a peculiar way.
For example, if the F1 score is 1, by the assumption that alpha is too small, there's no way to generate F score in (1 - alpha, 1] for __any__ assignment of labels across the validation set except when everything is correctly predicted. The effect of the validation loss is now mapped to the semi-open interval (0, alpha] by the mapping np.exp(-net.history[-1, 'valid_loss']) * alpha with equality on the right side of the interval iff validation loss is 0. Hence, smaller losses have higher total F1 value (for equal original F1 values). Moreover, this mapping works across the whole range of F1 values including the case of F1 values < alpha.

Thanks!
